### PR TITLE
Mysql source

### PR DESCRIPTION
**Pull request by danielgerlag**
This pull request adds a new MySQL bootstrap & source provider components to the project, enabling initial snapshotting of MySQL tables for query bootstrapping, and CDC.

These components were generated with the experimental source planner & executor agents from #170

### CODE REVIEW

File: components/sources/mysql/src/decoder.rs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+            }
+
+            properties.insert(&column_key, value);
+        }
+
+        if key_parts.is_empty() {
+            warn!("No primary key mapping for table '{table_name}', falling back to UUID");
+        }
+
+        let element_id = if !key_parts.is_empty() {
```
Comment (amansinghoriginal) [inline components/sources/mysql/src/decoder.rs]: Looks like Element IDs generated here will never match the bootstrapper, even with table_keys configured.

1. Table name prefix: This builds "{db}.{table}" (like mydb.users:1), but the bootstrapper uses the bare table name (users:1).

2. id fallback: CDC falls back to a column named id when no table_keys are configured. The bootstrapper has no such fallback and goes straight to UUID.

3. table_keys lookup: This looks up self.table_keys.get("mydb.users"), but the bootstrapper looks up self.table_keys.get("users").

The Postgres bootstrapper solves this by auto-discovering primary keys from system catalogs and using a consistent table name format.

File: components/sources/mysql/src/stream.rs [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+            }
+
+            if let Some(state_store) = state_store.as_deref() {
+                if let Err(e) = self.persist_state(state_store, &header).await {
+                    warn!("Failed to persist replication state: {e:?}");
+                }
+            }
+        }
+
+        Ok(())
```
Comment (amansinghoriginal) [inline components/sources/mysql/src/stream.rs:272]: This `Ok(())` silently masks connection failure. The source dies but reports as healthy.

Am I understanding this correctly - if the binlog connection drops, the spawn_blocking thread breaks out of its loop, the channel sender drops, the while let loop ends, and this line returns Ok?

Comment (amansinghoriginal) [reply components/sources/mysql/src/stream.rs:272]: Do we need a Reconnection loop for MySQL as well? PostgreSQL source has recover_connection method that closes the old connection, waits, and reconnects.

And maybe here we should return Err when the channel closes unexpectedly. Then lib.rs can at least update the status to Error.

Comment (amansinghoriginal) [reply components/sources/mysql/src/stream.rs:272]: A channel closing without a shutdown signal is always an error condition. Right?

File: components/sources/mysql/src/decoder.rs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+            format!("{}:{}", table_name, key_parts.join("_"))
+        } else {
+            format!("{}:{}", table_name, uuid::Uuid::new_v4())
+        };
+
+        properties.insert(
+            "_element_id",
+            ElementValue::String(Arc::from(element_id.as_str())),
```
Comment (amansinghoriginal) [inline components/sources/mysql/src/decoder.rs]: We don't add this field for postgres source. Is this important for queries/reactions?

File: components/sources/mysql/src/stream.rs [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+                }
+            }
+        });
+
+        while let Some((header, event)) = event_rx.recv().await {
+            if let Err(e) = self.process_event(&header, &event).await {
+                error!("Error processing binlog event: {e:?}");
+            }
+
+            if let Some(state_store) = state_store.as_deref() {
```
Comment (amansinghoriginal) [inline components/sources/mysql/src/stream.rs]: State is persisted on every binlog event, including internal events and events for tables we don't monitor. This is a state store write per event could be a lot of I/O.

Should we flush every so often like the Postgres source does this on a 10-second timer, or maybe save it at transaction commit boundaries?

File: components/sources/mysql/src/decoder.rs [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        MySqlValue::Decimal(v) => v
+            .parse::<f64>()
+            .map(OrderedFloat)
+            .map(ElementValue::Float)
+            .unwrap_or_else(|_| ElementValue::String(Arc::from(v.as_str()))),
+        MySqlValue::String(v) => ElementValue::String(Arc::from(v.as_str())),
+        MySqlValue::Bit(v) => {
+            ElementValue::List(v.iter().map(|b| ElementValue::Bool(*b)).collect::<Vec<_>>())
+        }
+        MySqlValue::Enum(v) => ElementValue::Integer(*v as i64),
```
Comment (amansinghoriginal) [inline components/sources/mysql/src/decoder.rs]: Enum, Set, and Timestamp mappings differ from the bootstrapper. CDC is returning numeric values (index/epoch) while bootstrap is returning strings.

Is this an issue? Will they not be translated to different Drasi types for same data?

**Review by amansinghoriginal — CHANGES_REQUESTED** [review]
Overall structure looks good.

However, there are two main issues that need to be addressed before merge:

1. Element IDs mismatch between bootstrap and CDC
2. The source silently dies on connection loss and reports as healthy

There might seems to be type mapping inconsistency between streaming mode and bootstrap mode.

File: components/sources/mysql/tests/integration_test.rs
Comment (amansinghoriginal) [inline components/sources/mysql/tests/integration_test.rs:1]: Test coverage is thin compared to the Postgres source. Consider adding
- a connectivity/lifecycle test
- separate tests per operation instead of all of them sequentially
- a failure-path test (like connection that reports Error status).

File: components/sources/mysql/src/decoder.rs (hunk 1 above)
Comment (danielgerlag) [reply components/sources/mysql/src/decoder.rs]: good catch

File: components/sources/mysql/src/stream.rs (hunk 2 above)
Comment (danielgerlag) [reply components/sources/mysql/src/stream.rs:272]: good catch

File: components/sources/mysql/src/decoder.rs (hunk 1 above)
Comment (danielgerlag) [reply components/sources/mysql/src/decoder.rs]: Fixed. Three changes made:

1. **Table name prefix removed** — CDC decoder now uses bare table name (`table_name`) instead of `db.table_name`, matching the bootstrapper.
2. **Unified key fallback** — Bootstrapper now has the same `id` column auto-detection fallback as the CDC decoder.
3. **table_keys lookup** — Both now use bare table name as the lookup key.

Also removed the `_element_id` synthetic property (see next comment).

File: components/sources/mysql/src/stream.rs (hunk 2 above)
Comment (danielgerlag) [reply components/sources/mysql/src/stream.rs:272]: Fixed. Three changes:

1. **Error on unexpected closure** — `run()` now returns `Err("Binlog replication channel closed unexpectedly")` when the channel closes without a shutdown signal. `lib.rs` catches this and sets `ComponentStatus::Error`.
2. **Reconnection loop** — Added retry loop with exponential backoff (5s → 60s, up to 10 attempts). Only returns fatal error after exhausting retries.
3. **Shutdown signal** — Added a shared `AtomicBool` flag. `stop()` sets it so `run()` can distinguish intentional shutdown from crash. The blocking thread also checks it to exit cleanly.

File: components/sources/mysql/src/decoder.rs (hunk 3 above)
Comment (danielgerlag) [reply components/sources/mysql/src/decoder.rs]: Removed. The `_element_id` synthetic property is no longer added — consistent with how the Postgres source works.

File: components/sources/mysql/src/stream.rs (hunk 4 above)
Comment (danielgerlag) [reply components/sources/mysql/src/stream.rs]: Fixed. State persistence moved from every-event to transaction commit boundaries (`XidEvent` / `COMMIT` query). The `flush_transaction()` method now handles both dispatching buffered changes and persisting state in one step.

File: components/sources/mysql/src/decoder.rs (hunk 5 above)
Comment (danielgerlag) [reply components/sources/mysql/src/decoder.rs]: Fixed. CDC decoder now maps consistently with the bootstrapper:
- `Enum(v)` → `String` (was `Integer`)
- `Set(v)` → `String` (was `Integer`)
- `Timestamp(ts)` → formatted datetime `String` (was `Integer` epoch)

Also added an integration test (`test_type_mapping_consistency_between_bootstrap_and_cdc`) that inserts a row with 16 diverse column types, bootstraps it, triggers a CDC update, and asserts all unchanged columns produce identical values from both paths.

File: components/sources/mysql/tests/integration_test.rs
Comment (danielgerlag) [reply components/sources/mysql/tests/integration_test.rs:1]: Added `test_type_mapping_consistency_between_bootstrap_and_cdc` — an end-to-end integration test covering 16 column types (INT, BIGINT, FLOAT, DOUBLE, DECIMAL, VARCHAR, TINYINT, BOOLEAN, DATE, TIME, DATETIME, TIMESTAMP, YEAR, ENUM, SET). It bootstraps a row, triggers a CDC update, and asserts all column values match between the two paths.

Will add the additional lifecycle and failure-path tests in a follow-up.

File: components/bootstrappers/mysql/src/mysql.rs [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        _settings: Option<&drasi_lib::config::SourceSubscriptionSettings>,
+    ) -> Result<usize> {
+        info!(
+            "Starting MySQL bootstrap for source {} and query {}",
+            context.source_id, request.query_id
+        );
+
+        self.config.validate()?;
+
+        let mut conn = self.connect().await?;
```
Comment (ruokun-niu) [inline components/bootstrappers/mysql/src/mysql.rs:80]: The MSSQL bootstrapper uses `SET TRANSACTION ISOLATION LEVEL SNAPSHOT` to ensure consistent
  reads during bootstrap. Should we consider adding something like `START TRANSACTION WITH CONSISTENT SNAPSHOT` here to prevent inconsistent state if writes occur during bootstrap.

**Review by ruokun-niu — APPROVED** [review]
Looks good. I think we need to add support for identity providers but this can probably be handled in another PR

File: components/sources/mysql/tests/integration_test.rs [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+use testcontainers::{runners::AsyncRunner, ImageExt};
+use testcontainers_modules::mysql::Mysql;
+use tokio::time::{sleep, timeout, Duration};
+
+use drasi_bootstrap_mysql::MySqlBootstrapProvider;
+use drasi_source_mysql::{MySqlReplicationSource, StartPosition, TableKeyConfig};
+
+#[tokio::test]
+#[ignore]
+async fn test_mysql_source_change_detection() {
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/tests/integration_test.rs:30]: Not applicable — #[ignore] on integration tests is standard Rust convention. Integration tests requiring external services (databases, containers, APIs) are run separately via cargo test -- --ignored. This is the expected pattern.

File: components/sources/mysql/src/decoder.rs [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+            if let Some(names) = &metadata.column_names {
+                return names.clone();
+            }
+        }
+        Vec::new()
+    }
+}
+
+fn mysql_value_to_element_value(value: &MySqlValue) -> ElementValue {
+    match value {
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/src/decoder.rs:184]: Addressed — Added unit tests for mysql_value_to_element_value covering all MySqlValue variants.

File: components/sources/mysql/src/decoder.rs [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        }
+
+        if key_parts.is_empty() {
+            warn!("No primary key mapping for table '{table_name}', falling back to UUID");
+        }
+
+        let element_id = if !key_parts.is_empty() {
+            format!("{}:{}", table_name, key_parts.join("_"))
+        } else {
+            format!("{}:{}", table_name, uuid::Uuid::new_v4())
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/src/decoder.rs]: Fixed: replaced the UUID fallback with `anyhow::bail!()` in both the source decoder and bootstrapper path, so we now return an actionable error when no deterministic key can be constructed.

File: components/sources/mysql/src/stream.rs [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+                }
+            };
+
+            for result in events {
+                if shutdown.load(Ordering::Relaxed) {
+                    break;
+                }
+                match result {
+                    Ok((header, event)) => {
+                        client.commit(&header, &event);
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/src/stream.rs]: Addressed: migrated CDC consumption from `mysql_cdc` to native async `mysql_async` BinlogStream, which removes the old blocking bridge and commit-ordering risk.

File: components/sources/mysql/src/stream.rs [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+    async fn run_replication_loop(
+        &mut self,
+        options: ReplicaOptions,
+        state_store: Option<&dyn drasi_lib::state_store::StateStoreProvider>,
+    ) -> Result<()> {
+        let (event_tx, mut event_rx) = tokio::sync::mpsc::channel(1000);
+        let source_id = self.source_id.clone();
+        let shutdown = self.shutdown.clone();
+
+        tokio::task::spawn_blocking(move || {
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/src/stream.rs]: Resolved: the `mysql_async` BinlogStream migration removes the `spawn_blocking` worker entirely, so there is no dropped `JoinHandle` or hidden panic path anymore.

File: components/bootstrappers/mysql/src/mysql.rs [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+        let columns = row.columns_ref();
+        for (idx, column) in columns.iter().enumerate() {
+            let col_name = column.name_str().to_string();
+            let value = match row.as_ref(idx) {
+                Some(mysql_async::Value::NULL) => ElementValue::Null,
+                Some(mysql_async::Value::Bytes(bytes)) => {
+                    ElementValue::String(Arc::from(String::from_utf8_lossy(bytes).as_ref()))
+                }
+                Some(mysql_async::Value::Int(val)) => ElementValue::Integer(*val),
+                Some(mysql_async::Value::UInt(val)) => ElementValue::Integer(*val as i64),
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/bootstrappers/mysql/src/mysql.rs]: Fixed: `UInt` values now convert to `i64` only when `val <= i64::MAX as u64`; larger values fall back to string representation instead of wrapping.

File: components/bootstrappers/mysql/src/config.rs [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+/// Table key configuration for MySQL bootstrap
+#[derive(Debug, Clone, PartialEq)]
+pub struct TableKeyConfig {
+    pub table: String,
+    pub key_columns: Vec<String>,
+}
+
+/// MySQL bootstrap provider configuration
+#[derive(Debug, Clone, PartialEq)]
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/bootstrappers/mysql/src/config.rs]: Resolved: added a manual `Debug` impl for the bootstrapper config that redacts the password field as `[REDACTED]`.

File: components/sources/mysql/src/config.rs [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    FromGtid(String),
+}
+
+impl Default for StartPosition {
+    fn default() -> Self {
+        Self::FromEnd
+    }
+}
+
+/// MySQL replication source configuration
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/src/config.rs:66]: Resolved: added a manual `Debug` impl for the source config that redacts the password field as `[REDACTED]`.

File: components/sources/mysql/src/config.rs [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+        }
+
+        if self.port == 0 {
+            return Err(anyhow::anyhow!(
+                "Validation error: port cannot be 0. \
+                 Please specify a valid port number (1-65535)"
+            ));
+        }
+
+        if self.ssl_mode != SslMode::Disabled {
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/src/config.rs]: Addressed: migrated to native async `mysql_async` binlog streaming, which supports SSL, and changed the default `SslMode` to `IfAvailable`.

File: components/sources/mysql/src/stream.rs [hunk 16]
```diff
@@ -0,0 +1,432 @@
...
+
+    fn build_replica_options(&self, start_position: StartPosition) -> Result<ReplicaOptions> {
+        let ssl_mode = match self.config.ssl_mode {
+            SslMode::Disabled => MySqlCdcSslMode::Disabled,
+            SslMode::IfAvailable => MySqlCdcSslMode::IfAvailable,
+            SslMode::Require => MySqlCdcSslMode::Require,
+            SslMode::RequireVerifyCa => MySqlCdcSslMode::RequireVerifyCa,
+            SslMode::RequireVerifyFull => MySqlCdcSslMode::RequireVerifyFull,
+        };
+
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/src/stream.rs:577]: Fixed: when the `tables` allowlist is empty we now emit a `warn!()` so the replicate-all behavior is explicit in logs.

File: components/bootstrappers/mysql/src/config.rs [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+// limitations under the License.
+
+//! Configuration types for the MySQL bootstrap provider.
+//!
+//! These types are defined locally to keep this component independent
+//! and self-contained, without dependencies on other components.
+
+/// Table key configuration for MySQL bootstrap
+#[derive(Debug, Clone, PartialEq)]
+pub struct TableKeyConfig {
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/bootstrappers/mysql/src/config.rs]: Resolved: extracted the shared MySQL keying and identifier helpers into a new `mysql-common` crate, including `TableKeyConfig`.

File: components/sources/mysql/src/config.rs [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+//! This source monitors MySQL databases using binlog replication to stream
+//! data changes as they occur.
+
+// =============================================================================
+// SSL Configuration
+// =============================================================================
+
+/// SSL mode for MySQL connections
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum SslMode {
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/src/config.rs:28]: Addressed: updated `SslMode` for the `mysql_async` implementation, removed the old `mysql_cdc`-specific restriction, and set the default to `IfAvailable`.

File: components/sources/mysql/src/decoder.rs [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+                Some(dt) => {
+                    ElementValue::String(Arc::from(dt.format("%Y-%m-%d %H:%M:%S.000").to_string()))
+                }
+                None => ElementValue::Integer(secs),
+            }
+        }
+    }
+}
+
+fn format_value_for_key(value: &ElementValue) -> String {
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/src/decoder.rs]: Resolved: `format_value_for_key` now lives in the shared `mysql-common` crate alongside the other shared identifier and key helpers.

File: components/sources/mysql/tests/integration_test.rs [hunk 20]
```diff
@@ -0,0 +1,461 @@
...
+    let root_opts = mysql_async::OptsBuilder::default()
+        .ip_or_hostname("127.0.0.1")
+        .tcp_port(port)
+        .user(Some("root"))
+        .pass(Some("root"))
+        .db_name(Some("test"));
+
+    let mut conn = Conn::new(root_opts).await.unwrap();
+    sleep(Duration::from_secs(2)).await;
+
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/tests/integration_test.rs:66]: Fixed: replaced the fixed `sleep()` calls in the integration tests with retry loops so the assertions wait for readiness without hard-coded timing.

File: components/sources/mysql/src/tests.rs [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+            ssl_mode: SslMode::Disabled,
+            table_keys: vec![],
+            start_position: StartPosition::FromEnd,
+            server_id: 65535,
+            heartbeat_interval_seconds: 30,
+        };
+        assert!(config.validate().is_err());
+    }
+
+    #[test]
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/src/tests.rs:35]: Addressed: added 9 config validation tests covering the missing edge cases.

File: components/bootstrappers/mysql/README.md [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+
+```bash
+cargo test -p drasi-bootstrap-mysql
+```
+
+## Security
+
+- Configure tables explicitly with `with_tables`; this allowlist is required.
+- Table names must use only letters, numbers, and underscores.
+- Requested labels not in the allowlist are ignored.
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/bootstrappers/mysql/README.md]: Fixed: updated the bootstrapper README to use `tables` instead of `labels`.

File: components/bootstrappers/mysql/README.md [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+## Configuration Options
+
+| Property | Type | Default | Description |
+|----------|------|---------|-------------|
+| `host` | `String` | `"localhost"` | MySQL server hostname or IP address |
+| `port` | `u16` | `3306` | MySQL server port number |
+| `database` | `String` | **(Required)** | Database name to connect to |
+| `user` | `String` | **(Required)** | Database user with replication privileges |
+| `password` | `String` | `""` | Database password |
+| `tables` | `Vec<String>` | `[]` | Table allow-list (required for security) |
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/bootstrappers/mysql/README.md]: Fixed: corrected the bootstrapper README so `tables` is documented as required rather than implying a default.

File: components/sources/mysql/README.md [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+- `mysql_cdc` runtime does not support SSL (ssl_mode must be Disabled). This means connections from this source to MySQL are **not encrypted**.
+  - **Do not** use this source over untrusted networks or the public internet.
+  - In production, run it only on trusted, isolated networks and ensure encryption in transit via infrastructure such as a VPN, service mesh, or a TLS-terminating proxy (for example, stunnel, HAProxy, or a cloud load balancer that terminates MySQL TLS and forwards plain traffic on a private network).
+- Packets >16MB are not supported
+
+## Testing
+
+Integration test uses testcontainers:
+
+```bash
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/README.md:60]: Resolved: cleaned up the source README formatting and clarified the `serverId` description.

File: components/sources/mysql/Cargo.toml [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+log = "0.4"
+tokio = { version = "1.0", features = ["full"] }
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0"
+tracing = "0.1"
+chrono = { version = "0.4", features = ["serde"] }
+ordered-float = "3.0"
+uuid = { version = "1.0", features = ["v4", "serde"] }
+mysql_cdc = "0.2.1"
+bytes = "1.0"
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/mysql/Cargo.toml:50]: Addressed: replaced `mysql_cdc` with native async `mysql_async` binlog streaming, which removes the blocking bridge and related ordering and panic issues.

**Review by agentofreality — APPROVED** [review]
